### ISSUE TITLE
repeated calls to fill() slow in webgl mode

### ISSUE THREAD

**Issue by kjhollen** [processing/p5.js#2190]
#### Nature of issue?

- [x] Found a bug
- [ ] Existing feature enhancement
- [ ] New feature request

#### Most appropriate sub-area of p5.js?

- [ ] Color
- [ ] Core
- [ ] Data
- [ ] Events
- [ ] Image
- [ ] IO
- [ ] Math
- [ ] Typography
- [ ] Utilities
- [x] WebGL
- [x] Friendly Error System?

#### Which platform were you using when you encountered this?

- [ ] Mobile/Tablet (touch devices)
- [x] Desktop/Laptop
- [ ] Others (specify if possible)

#### Details about the bug: 

- p5.js version: (as yet unreleased development branch)
- Web browser and version: 60.0.3112.113
- Operating System: macOS Sierra
- Steps to reproduce this:

this manual test snippet (in [test/manual-test-examples/webgl/immediateMode/performance](https://github.com/processing/p5.js/tree/master/test/manual-test-examples/webgl/immediateMode/performance)) took a performance hit when we merged the web gl summer of code branch with the master branch. the snippet draws a bunch of random magenta triangles, each tinted slightly differently.

```
function setup(){
  createCanvas(windowWidth, windowHeight, WEBGL);
}

function draw(){
  background(250);

  randomSeed(1);
  noStroke();
  for (var i = 0  ;  i < 1000 ;  i++) {
    push();
    translate(random(-width/2, width/2), random(-height/2, height/2));
    rotateZ(random(0, 2 * PI) + frameCount * 0.1);
    fill(random(128, 255), 0, random(128, 255));
    beginShape(TRIANGLES);
    vertex(0, 25, 0);
    vertex(-25, -25, 0);
    vertex(25, -25, 0);
    endShape();
    pop();
  }
}
```
@mlarghydracept this is what we saw the other day. it looks like the slow step is validating the arguments to fill(), rather than the gl code itself (see attached performance profile). @almchung any thoughts?

![screen shot 2017-09-19 at 12 23 00 pm](https://user-images.githubusercontent.com/78966/30673903-421e4516-9e2b-11e7-8344-7ecd20000fe9.png)

**Comment by meiamsome** [comment]
This is really interesting, actually, because this performs better (about a 30% performance bump on my computer) after having the #2177 PR applied to it.

The main reason for the slowness is that the `validateParameters` function currently has to perform a lookup on the entire documentation array (O(n)) rather than having a reference to the correct documentation as is possible when FES is injected directly.

Looking bottom up we see most of the time is in the anonymous function used to filter the documentation:
![image](https://user-images.githubusercontent.com/1187491/30674335-66385ff4-9e71-11e7-8090-81121d46d2c6.png)

![image](https://user-images.githubusercontent.com/1187491/30674360-84b2f6a6-9e71-11e7-949d-4bf0abffc0a9.png)

With FES injected we instead see:
![image](https://user-images.githubusercontent.com/1187491/30674445-0d9ede6c-9e72-11e7-91e2-5e5a3b551dcc.png)

**Comment by almchung** [comment]
@kjhollen hey Kate, this is probably due to FES's (slow) lookup function as Oliver mentioned above. Clearer diagnosis can be derived via running the code with p5.js.min instead of p5.js (FES disabled).

**Comment by Spongman** [comment]
this was fixed with #2302

**Comment by mlarghydracept** [comment]
Thanks @Spongman !
